README.md
TLS Interposer
==============

OpenSSL library interposer to get software to use more secure TLS protocol variants.

Functionality
-------------
1. Disables SSLv2 and SSLv3 (broken), enables everything else (starting at OpenSSL 1.0.1 up to TLS 1.2)
2. Enables ECDHE algorithms for forward secrecy
3. Disables all weak algorithms; by default including RC4 as a last resort compatibility mode

Environment Variables
---------------------
* `LD_PRELOAD`: Used by ld.so, should be set to /full/path/to/tlsinterposer.so
* `TLS_INTERPOSER_CIPHERS`: The ciphers to use, defaults to Qualys SSLlabs recommendations
* `TLS_INTERPOSER_OPTIONS`: Comma-separated list of options
  - `debug`: Be verbose, by default on stderr
  - `logfile`: Log to /var/log/tlsinterposer.log; fall back to stderr
  - `ssllib=`: Full name of libssl.so.X.Y.Z, if not autodetected correctly
  - `-comp`: Disable compression
  - `-ccert`: Prevent the server from asking for client certificates
  - `+sorder`: Force server cipher order, aka tls_preempt_cipherlist (Postfix) or SSLHonorCipherOrder (Apache httpd)
  - `-rc4`: Changes the default ciphers from Qualys recommendations with to without RC4 (has no effect on TLS_INTERPOSER_CIPHERS)
  - `-tlsv1`: Disable TLSv1, leaving TLSv1.1 and TLSv1.2, if supported
  - `+sslv3`: Reenable SSLv3 (advised against)
  - `+sslv2`: Reenable SSLv2 (strongly advised against)

Installation
------------
Download, make, make install.

Usage
-----
Start the process with LD_PRELOAD environment variable set to /path/to/libtlsinterposer.so . For example,

	env LD_PRELOAD=/usr/local/lib/libtlsinterposer.so apache2ctl start

enables Apache 2.2 to use the modern ciphers.

More information and documentation is available at https://netfuture.ch/tools/tls-interposer/


CHANGES.txt
Changes
=======

1.0.0   2013-11-19
        First "official" release with version and license

1.1.0   2013-12-12
        Standardized DH parameters based on the length of the certificate's RSA/DSA key

1.2.0   2014-06-26
        Refactored the code to improve readability
        Configuration through TLS_INTERPOSER_OPTIONS
        More consistent enabling of ECDHE
        More hooks used to tighten security

1.2.1   2014-06-27
        Add -rc4; eliminate most uses of TLS_INTERPOSER_CIPHERS
        Optionally log to /var/log/tlsinterposer.log

1.2.2   2014-07-04
        Eliminated support for deprecated TLS_INTERPOSER_NO_COMPRESSION
        Support OpenSSL 0.9.8* on non-Apple platforms

1.3.0   2014-07-05
        Removed (IMHO useless) -ecdhe option, reenabled ECDHE always
        More ciphers with OpenSSL 0.9.8
        New "-ccert" option
        Debug output now includes process name

1.3.1	2014-11-xx (unreleased)
        New option "-sorder" (honor server cipher order)
	More flexible openssl library version parsing
	


tests/simple_server.c
// based on serv.cpp example from the openssl distribution

/* serv.cpp    -    Minimal ssleay server for Unix
     30.9.1996, Sampo Kellomaki <sampo@iki.fi> */


/* mangled to work with SSLeay-0.9.0b and OpenSSL 0.9.2b
     Simplified to be even more minimal
     12/98 - 4/99 Wade Scholine <wades@mail.cybg.com> */

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <memory.h>
#include <libgen.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>

#include <openssl/rsa.h>
#include <openssl/crypto.h>
#include <openssl/x509.h>
#include <openssl/pem.h>
#include <openssl/ssl.h>
#include <openssl/err.h>


/* Make these what you want for cert & key files */
#define CERTF   "test-cert.pem"
#define KEYF    "test-cert.pem"


#define CHK_NULL(x) if ((x)==NULL) exit (1)
#define CHK_ERR(err,s) if ((err)==-1) { perror(s); exit(1); }
#define CHK_SSL(err) if ((err)==-1) { ERR_print_errors_fp(stderr); exit(2); }

int verify_callback(int i, X509_STORE_CTX *ctx)
{
    (void)(i), (void)(ctx); // get rid of "unused param" warnings
    return 1;
}

int main (int argc, char *argv[])
{
    int err;
    int bytes;
    int listen_sd;
    int sd;
    struct sockaddr_in sa_serv;
    struct sockaddr_in sa_cli;
    socklen_t client_len;
    SSL_CTX* ctx;
    SSL*         ssl;
    X509*        client_cert;
    char*        str;
    char         buf [4096];
    const SSL_METHOD *meth;
    unsigned int nport;

    /* parse options */
    if (argc != 2) {
        fprintf(stderr, "usage: %s port_number\n", argv[0]);
        exit(255);
    }
    nport = strtoul(argv[1], &str, 0);
    if (*str != '\0' || nport >= 0xffff) {
        fprintf(stderr, "error parsing port number: %s\n", argv[1]);
        exit(255);
    }

    /* chdir to our directory, to load cert files properly if run from another dir */
    strncpy(buf, argv[0], sizeof(buf)); // copy since dirname may modify
    str = dirname(buf);
    err = chdir(str);
    CHK_ERR(err, "chdir");

    /* SSL preliminaries. We keep the certificate and key with the context. */

    SSL_load_error_strings();
    SSL_library_init();
    meth = SSLv23_server_method();
    ctx = SSL_CTX_new (meth);
    // an explicit setting to test for
    SSL_CTX_set_cipher_list(ctx, "AES128-SHA");
    if (!ctx) {
        ERR_print_errors_fp(stderr);
        exit(2);
    }

    if (SSL_CTX_use_certificate_file(ctx, CERTF, SSL_FILETYPE_PEM) <= 0) {
        ERR_print_errors_fp(stderr);
        exit(3);
    }
    if (SSL_CTX_use_PrivateKey_file(ctx, KEYF, SSL_FILETYPE_PEM) <= 0) {
        ERR_print_errors_fp(stderr);
        exit(4);
    }

    if (!SSL_CTX_check_private_key(ctx)) {
        fprintf(stderr,"Private key does not match the certificate public key\n");
        exit(5);
    }

    /* ----------------------------------------------- */
    /* Prepare TCP socket for receiving connections */

    listen_sd = socket (AF_INET, SOCK_STREAM, 0);     CHK_ERR(listen_sd, "socket");
    err = 1;
    err = setsockopt(listen_sd, SOL_SOCKET, SO_REUSEADDR, &err, sizeof(err));    CHK_ERR(err, "setsockopt");

    memset (&sa_serv, '\0', sizeof(sa_serv));
    sa_serv.sin_family            = AF_INET;
    sa_serv.sin_addr.s_addr = htonl (INADDR_LOOPBACK);
    sa_serv.sin_port                = htons (nport);                    /* Server Port number */

    err = bind(listen_sd, (struct sockaddr*) &sa_serv,
             sizeof (sa_serv));                                     CHK_ERR(err, "bind");

    /* Receive a TCP connection. */

    err = listen (listen_sd, 5);                                        CHK_ERR(err, "listen");

    client_len = sizeof(sa_cli);
    sd = accept (listen_sd, (struct sockaddr*) &sa_cli, &client_len);
    CHK_ERR(sd, "accept");
    close (listen_sd);

    printf ("Connection from %lx, port %x\n",
        (long unsigned int)sa_cli.sin_addr.s_addr, sa_cli.sin_port);

    /* ----------------------------------------------- */
    /* TCP connection is ready. Do server side SSL. */

    ssl = SSL_new (ctx);                                                     CHK_NULL(ssl);
    SSL_set_fd (ssl, sd);
    SSL_set_verify(ssl, SSL_VERIFY_PEER, verify_callback); /* ask for a cert */
    err = SSL_accept (ssl);                                                CHK_SSL(err);

    /* Get the cipher - opt */

    printf ("SSL connection using %s\n", SSL_get_cipher (ssl));

    /* Get client's certificate (note: beware of dynamic allocation) - opt */

    client_cert = SSL_get_peer_certificate (ssl);
    if (client_cert != NULL) {
        printf ("Client certificate:\n");

        str = X509_NAME_oneline (X509_get_subject_name (client_cert), 0, 0);
        CHK_NULL(str);
        printf ("\t subject: %s\n", str);
        err = SSL_write (ssl, "client cert: ", strlen("client cert: "));    CHK_SSL(err);
        err = SSL_write (ssl, str, strlen(str));    CHK_SSL(err);
        err = SSL_write (ssl, "\n", 1);        CHK_SSL(err);
        OPENSSL_free (str);

        str = X509_NAME_oneline (X509_get_issuer_name    (client_cert), 0, 0);
        CHK_NULL(str);
        printf ("\t issuer: %s\n", str);
        OPENSSL_free (str);

        /* We could do all sorts of certificate verification stuff here before
             deallocating the certificate. */

        X509_free (client_cert);
    } else
        printf ("Client does not have certificate.\n");

    /* DATA EXCHANGE - Receive message and send reply. */

    bytes = SSL_read (ssl, buf, sizeof(buf) - 1);                                     CHK_SSL(err);
    buf[bytes] = '\0';
    printf ("Got %d chars:'%s'\n", bytes, buf);

    err = SSL_write (ssl, "echo: ", strlen("echo: "));    CHK_SSL(err);
    err = SSL_write (ssl, buf, bytes);    CHK_SSL(err);

    /* Clean up. */

    close (sd);
    SSL_free (ssl);
    SSL_CTX_free (ctx);

    return 0;
}
/* EOF - serv.cpp */


tlsinterposer.c
/*
    TLS Interposer: An OpenSSL library interposer to get
    software to use more secure TLS protocol variants.

    Copyright (C) 2013,2014 Marcel Waldvogel

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
    USA
*/
#define _GNU_SOURCE
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <time.h>
#include <openssl/ssl.h>
#include <openssl/dh.h>
#include <dlfcn.h>
#include "ssl-version.h"

/* Environment variables used
 * ==========================
 * LD_PRELOAD               used by ld.so, should be set to /full/path/to/tlsinterposer.so
 * TLS_INTERPOSER_CIPHERS   defaults to DEFAULT_CIPHERS below
 * TLS_INTERPOSER_OPTIONS   comma-separated list of options
 * - debug                  be verbose
 * - logfile                log to /var/log/tlsinterposer.log instead of stderr
 * - ssllib=                full name of libssl.so.X.Y.Z
 * - -comp                  disable compression
 * - -rc4                   remove RC4 from default (!) ciphers
 * - -ccert                 disable client certificate requests on the server side
 * - +sorder                honor server cipher order
 * - +sslv2                 enable SSLv2 (strongly advised against)
 * - +sslv3                 enable SSLv3 (advised against)
 * - -tlsv1                 disable TLSv1, leaving TLSv1.1 and TLSv1.2, if supported
*/

#ifdef SSL_OP_NO_COMPRESSION // OpenSSL 1.0.0 or newer?
// Qualys recommendation (I know the RC4 part could be simplified)
// - https://community.qualys.com/blogs/securitylabs/2013/08/05/configuring-apache-nginx-and-openssl-for-forward-secrecy
#define DEFAULT_CIPHERS "EECDH+ECDSA+AESGCM EECDH+aRSA+AESGCM EECDH+ECDSA+SHA384 EECDH+ECDSA+SHA256 EECDH+aRSA+SHA384 EECDH+aRSA+SHA256 EECDH+aRSA+RC4 EECDH EDH+aRSA RC4 !aNULL !eNULL !LOW !3DES !MD5 !EXP !PSK !SRP !DSS +RC4 RC4"
#define CIPHERS_NO_RC4 "EECDH+ECDSA+AESGCM EECDH+aRSA+AESGCM EECDH+ECDSA+SHA384 EECDH+ECDSA+SHA256 EECDH+aRSA+SHA384 EECDH+aRSA+SHA256 EECDH+aRSA+RC4 EECDH EDH+aRSA RC4 !aNULL !eNULL !LOW !3DES !MD5 !EXP !PSK !SRP !DSS +RC4 RC4 !RC4"
#define SSLCONST const
#else
// The OpenSSL 0.9.8 family only supports RC4-SHA with the above cipher lists
// This prevents connecting to many servers which do not accept RC4 at all
#define DEFAULT_CIPHERS "DHE-RSA-AES256-SHA AES256-SHA RC4-SHA"
#define CIPHERS_NO_RC4 "DHE-RSA-AES256-SHA AES256-SHA"
#define SSLCONST
#endif

#define LOGFILE "/var/log/tlsinterposer.log"
#define ERRORLOG(...) interposer_log(__VA_ARGS__)
#ifdef NDEBUG
#define DEBUGLOG(...)
#else
#define DEBUGLOG(...) do {if (interposer_debug != 0) interposer_log(__VA_ARGS__);} while (0)
#endif

// interposer_debug, used by DEBUGLOG(), is only valid after the first call to interposer_dlsym()
#define ORIG_FUNC(func, rettype, args, fail)                    \
    static rettype (*orig_ ## func) args;                    \
    if (orig_ ## func == NULL) orig_ ## func = interposer_dlsym( #func );    \
    DEBUGLOG("Intercepted call to %s\n", __func__);            \
    if (orig_ ## func == NULL) return fail;

static int   interposer_inited     = 0,
             interposer_opt_set    = (SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3 | SSL_OP_SINGLE_DH_USE),
             interposer_opt_clr    = 0,
             interposer_debug      = 0,
             interposer_tofile     = 0,
             interposer_no_ccert   = 0;
static char *interposer_ssllib     = DEFAULT_SSLLIB,
            *interposer_ciphers    = DEFAULT_CIPHERS;

void interposer_log(const char *format, ...) __attribute__ ((format (printf, 1, 2)));
void interposer_log(const char *format, ...)
{
    static char buf[1024];
    static const char *progname = NULL;
    FILE *log = stderr;
    va_list ap;
    time_t t;
    struct tm tm;

    // Try to obtain command name from /proc once; fall back to "?"
    if (progname == NULL) {
        if (readlink("/proc/self/exe", buf, sizeof(buf)-1) > 0) {
            buf[sizeof(buf)-1] = '\0';
            const char *tail = strrchr(buf, '/');
            if (tail == NULL) {
                progname = buf;
            } else {
                progname = tail+1;
            }
        } else {
            progname = "?";
        }
    }

    t = time(NULL);
    localtime_r(&t, &tm);
    // Try to interfere as little as possible with the user program's fds
    if (interposer_tofile != 0) log = fopen(LOGFILE, "a");
    if (log == NULL) log = stderr;
    va_start(ap, format);
    // Fall back to stderr on problems
    fprintf(log, "%04d-%02d-%02d %02d:%02d:%02d tlsinterposer[%d]: ",
        1900 + tm.tm_year, 1 + tm.tm_mon, tm.tm_mday, tm.tm_hour, tm.tm_min, tm.tm_sec, getpid());
    vfprintf(log, format, ap);
    va_end(ap);
    if (log != stderr) fclose(log);
}

static void interposer_parse_opts(void)
{
    char *opts, *optend;
    size_t optlen;
    char *ciphers;

    // This is only needed to improve efficiency, not correctness,
    // so an at-least-once semantic is used (increment at the end of the function)
    if (interposer_inited > 0) return;

    opts = getenv("TLS_INTERPOSER_OPTIONS");
    if (opts == NULL) opts = "";
    /* Non-destructive strtok() clone */
    while (*opts != '\0') {
        optend = index(opts, ',');
        if (optend == NULL) {
            optlen = strlen(opts); // Until EOS
            optend = opts + optlen;
        } else {
            optlen = optend - opts;
            optend++;
        }
        if (strncasecmp(opts, "debug", optlen) == 0) {
            interposer_debug++;
        } else if (strncasecmp(opts, "logfile", optlen) == 0) {
            interposer_tofile++;
        } else if (strncasecmp(opts, "+sslv2", optlen) == 0) {
            interposer_opt_set &= ~SSL_OP_NO_SSLv2;
            interposer_opt_clr |= SSL_OP_NO_SSLv2;
        } else if (strncasecmp(opts, "+sslv3", optlen) == 0) {
            interposer_opt_set &= ~SSL_OP_NO_SSLv3;
            interposer_opt_clr |= SSL_OP_NO_SSLv3;
        } else if (strncasecmp(opts, "-tlsv1", optlen) == 0) {
            interposer_opt_set |= SSL_OP_NO_TLSv1;
#ifdef SSL_OP_NO_COMPRESSION
        } else if (strncasecmp(opts, "-comp", optlen) == 0) {
            interposer_opt_set |= SSL_OP_NO_COMPRESSION;
#endif
        } else if (strncasecmp(opts, "-ccert", optlen) == 0) {
            interposer_no_ccert++;
        } else if (strncasecmp(opts, "+sorder", optlen) == 0) {
            interposer_opt_set |= SSL_OP_CIPHER_SERVER_PREFERENCE;
        } else if (strncasecmp(opts, "-rc4", optlen) == 0) {
            interposer_ciphers = CIPHERS_NO_RC4;
        } else if (optlen > 7 && strncasecmp(opts, "libssl=", 7) == 0) {
            interposer_ssllib = opts+7;
        } else if (interposer_debug) {
            fprintf(stderr, "tlsinterposer.so: WARNING: Unknown option '%.*s' found in TLS_INTERPOSER_OPTIONS\n", (int)optlen, opts);
        }
        opts = optend;
    }
    // Higher priority than -rc4 above
    ciphers = getenv("TLS_INTERPOSER_CIPHERS");
    if (ciphers != NULL) interposer_ciphers = ciphers;

    interposer_inited++;
}

// Get a symbol from libssl
static void *interposer_dlsym(const char *name)
{
    if (interposer_inited == 0) interposer_parse_opts();
    void *addr = dlsym(RTLD_NEXT, name);
    if (addr == NULL) {
        // Try again with a more specific name
        // Needed for ejabberd
        void *file = dlopen(interposer_ssllib, RTLD_LAZY | RTLD_GLOBAL | RTLD_NOLOAD);
        if (file != NULL) {
            addr = dlsym(file, name);
            if (addr == NULL) {
                ERRORLOG("Cannot find symbol %s in libssl %s\n", name, interposer_ssllib);
            }
            // Can be dlclose()d here. As it wasn't loaded (RTLD_NOLOAD), it won't be unloaded here
            dlclose(file);
        } else {
            ERRORLOG("Cannot find mapped libssl %s looking for symbol %s\n", interposer_ssllib, name);
        }
    }
    return addr;
}

// Make ciphers in interposer_ciphers sticky
int SSL_CTX_set_cipher_list(SSL_CTX *ctx, const char *str)
{
    ORIG_FUNC(SSL_CTX_set_cipher_list, int, (SSL_CTX *ctx, const char *str), 0);
    return (*orig_SSL_CTX_set_cipher_list)(ctx, interposer_ciphers);
}
int SSL_set_cipher_list(SSL *ssl, const char *str)
{
    ORIG_FUNC(SSL_set_cipher_list, int, (SSL *ssl, const char *str), 0);
    return (*orig_SSL_set_cipher_list)(ssl, interposer_ciphers);
}

// Make options in interposer_opt_{set,clr} sticky
// SSL_{CTX_,}{set,clear}_options() are actually macros which map to SSL_{CTX_,}ctrl()
long SSL_CTX_ctrl(SSL_CTX *ctx, int cmd, long larg, void *parg)
{
    ORIG_FUNC(SSL_CTX_ctrl, long, (SSL_CTX *, int, long, void *), 0);
    switch (cmd) {
    case SSL_CTRL_OPTIONS:
        larg = (larg | interposer_opt_set) & ~interposer_opt_clr;
        break;
    case SSL_CTRL_CLEAR_OPTIONS:
        larg = (larg & ~interposer_opt_set) | interposer_opt_clr;
        break;
    }
    return (*orig_SSL_CTX_ctrl)(ctx, cmd, larg, parg);
}
long SSL_ctrl(SSL *ssl, int cmd, long larg, void *parg)
{
    ORIG_FUNC(SSL_ctrl, long, (SSL *, int, long, void *), 0);
    switch (cmd) {
    case SSL_CTRL_OPTIONS:
        larg = (larg | interposer_opt_set) & ~interposer_opt_clr;
        break;
    case SSL_CTRL_CLEAR_OPTIONS:
        larg = (larg & ~interposer_opt_set) | interposer_opt_clr;
        break;
    }
    return (*orig_SSL_ctrl)(ssl, cmd, larg, parg);
}

// Based on Apache's bug #49559 by Kaspar Brand
// - https://issues.apache.org/bugzilla/show_bug.cgi?id=49559#c13
/*
 * Grab well-defined DH parameters from OpenSSL, see <openssl/bn.h>
 * (get_rfc*) for all available primes.
 */
#define make_get_dh(rfc,size,gen) \
static DH *get_dh##size(void) \
{ \
    DH *dh; \
    if (!(dh = DH_new())) { \
        return NULL; \
    } \
    dh->p = get_##rfc##_prime_##size(NULL); \
    BN_dec2bn(&dh->g, #gen); \
    if (!dh->p || !dh->g) { \
        DH_free(dh); \
        return NULL; \
    } \
    return dh; \
}

/*
 * Prepare DH parameters from 1024 to 4096 bits, in 1024-bit increments
 */
make_get_dh(rfc2409, 1024, 2)
make_get_dh(rfc3526, 2048, 2)
make_get_dh(rfc3526, 3072, 2)
make_get_dh(rfc3526, 4096, 2)

static DH *ssl_callback_TmpDH(SSL *ssl, int export, int keylen)
{
    EVP_PKEY *pkey = SSL_get_privatekey(ssl);
    int type = pkey ? EVP_PKEY_type(pkey->type) : EVP_PKEY_NONE;
    if ((type == EVP_PKEY_RSA) || (type == EVP_PKEY_DSA)) {
        keylen = EVP_PKEY_bits(pkey);
    }
    if (keylen >= 4096)
        return get_dh4096();
    else if (keylen >= 3072)
        return get_dh3072();
    else if (keylen >= 2048)
        return get_dh2048();
    else
        return get_dh1024();
}

void SSL_CTX_set_tmp_dh_callback(SSL_CTX *ctx, DH *(*tmp_dh_callback)(SSL *ssl, int is_export, int keylength))
{
    ORIG_FUNC(SSL_CTX_set_tmp_dh_callback, void, (SSL_CTX *, DH *(*)(SSL *, int, int)), /*void*/);
    (*orig_SSL_CTX_set_tmp_dh_callback)(ctx, ssl_callback_TmpDH);
}

void SSL_set_tmp_dh_callback(SSL *ssl, DH *(*tmp_dh_callback)(SSL *ssl, int is_export, int keylength))
{
    ORIG_FUNC(SSL_set_tmp_dh_callback, void, (SSL *, DH *(*)(SSL *, int, int)), /*void*/);
    (*orig_SSL_set_tmp_dh_callback)(ssl, ssl_callback_TmpDH);
}
// END Apache-bug derived code

/* ========================== Handling most options and ciphers */

SSL_CTX *SSL_CTX_new(SSLCONST SSL_METHOD *method)
{
    ORIG_FUNC(SSL_CTX_new, SSL_CTX *, (SSLCONST SSL_METHOD *), NULL);
    SSL_CTX *ctx = (*orig_SSL_CTX_new)(method);
    if (ctx != NULL) {
        SSL_CTX_set_options(ctx, interposer_opt_set);
        SSL_CTX_clear_options(ctx, interposer_opt_clr);
        SSL_CTX_set_cipher_list(ctx, interposer_ciphers);
        SSL_CTX_set_tmp_dh_callback(ctx, ssl_callback_TmpDH);
        // Based on code by Vincent Bernat
        // - http://vincent.bernat.im/en/blog/2011-ssl-perfect-forward-secrecy.html
        // - https://github.com/bumptech/stud/pull/61
#ifdef NID_X9_62_prime256v1
        EC_KEY *ecdh;
        ecdh = EC_KEY_new_by_curve_name(NID_X9_62_prime256v1);
        SSL_CTX_set_tmp_ecdh(ctx, ecdh);
        EC_KEY_free(ecdh);
        DEBUGLOG("ECDH Initialized with NIST P-256\n");
#endif
    }
    DEBUGLOG("SSL_CTX_new returning %p\n", ctx);
    return ctx;
}

/* ========================== Handling "-ccert" */

/* When intercepting SSL_CTX_set_verify(), two modes of server operations possible:
 * - SSL_CTX does not have accept/connect information, it is set only after SSL_new():
 *   SSL_set_accept_state() and SSL_set_verify() need to be intercepted anyway
 * - SSL_CTX has accept/connect, SSL_new() "clones" from SSL_CTX:
 *   SSL_set_accept_state() and SSL_set_verify() may not be called,
 *   everything is handled in SSL_new()
 * So there is no need to intercept SSL_CTX_verify(), especially as
 * SSL_is_server() might be changed later, which would result in data loss
 */
void SSL_set_verify(SSL *s, int mode,
                    int (*verify_callback)(int, X509_STORE_CTX *))
{
    ORIG_FUNC(SSL_set_verify, void, (SSL *, int, int (*)(int, X509_STORE_CTX *)), /*void*/);
    if (interposer_no_ccert != 0 && s->server != 0) {
        // Disable requesting the client certificate
        mode = SSL_VERIFY_NONE;
    }
    (*orig_SSL_set_verify)(s, mode, verify_callback);
}

void SSL_set_accept_state(SSL *s)
{
    ORIG_FUNC(SSL_set_accept_state, void, (SSL *), /*void*/);
    (*orig_SSL_set_accept_state)(s);
    if (interposer_no_ccert) {
        SSL_set_verify(s, SSL_VERIFY_NONE, NULL); // NULL indicates no change, phew!
    }
}

SSL *SSL_new(SSL_CTX *ctx)
{
    ORIG_FUNC(SSL_new, SSL *, (SSL_CTX *), NULL);
    SSL *s = (*orig_SSL_new)(ctx);
    if (s != NULL && interposer_no_ccert != 0 && s->server != 0) {
        SSL_set_verify(s, SSL_VERIFY_NONE, NULL); // NULL indicates no change, phew!
    }
    return s;
}
